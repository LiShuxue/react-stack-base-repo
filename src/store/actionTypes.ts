export const ACTION_TYPE_A: string = 'ACTION_TYPE_A';
export const ACTION_TYPE_B: string = 'ACTION_TYPE_B';
