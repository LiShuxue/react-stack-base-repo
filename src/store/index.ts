import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import reducer from './reducers';

// preload state object key must be same as reducer
const preloadState = {
  stateA: 'this is preload A',
  stateB: 'this is preload B'
};

// we not have to pass the preloadState, then it will use the default value in reducer as the state
const store = createStore(reducer, preloadState, applyMiddleware(thunk));

export default store;