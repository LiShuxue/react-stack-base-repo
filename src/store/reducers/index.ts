import { combineReducers } from 'redux';

import { handleActionA } from './moduleAReducer';
import { handleActionB } from './moduleBReducer';

// preload state object key must be same as reducer
const reducer = combineReducers({
  stateA: handleActionA,
  stateB: handleActionB
})

export default reducer;