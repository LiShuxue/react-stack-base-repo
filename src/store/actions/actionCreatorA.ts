import * as types from '../actionTypes';

export interface InterfaceActionA {
  type: string;
  text: string;
}
export const actionCreatorA = (actionDetail: string): InterfaceActionA => {
  return {
    type: types.ACTION_TYPE_A,
    text: actionDetail
  };
};
