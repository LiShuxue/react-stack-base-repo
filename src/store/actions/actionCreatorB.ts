import * as types from '../actionTypes';
import { Dispatch } from 'redux';
import { testGetMethod } from 'src/http/api';

const requestSuccess = () => {
  return {
    type: types.REQUEST_SUCCESS,
    data: 'request success'
  };
};
const requestFailed = () => {
  return {
    type: types.REQUEST_FAILED,
    data: 'request failed'
  };
};

export const asyncActionCreatorB = () => {
  return async (dispatch: Dispatch, getState: Function) => {
    console.log(getState());
    try {
      await testGetMethod();
      dispatch(requestSuccess());
    } catch (error) {
      dispatch(requestFailed());
    }
  };
};
