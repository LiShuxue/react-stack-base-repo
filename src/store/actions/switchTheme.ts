import * as types from '../actionTypes';

export const switchTheme = (theme: string): ActionType => {
  return {
    type: types.SWITCH_THEME,
    data: theme
  };
};
