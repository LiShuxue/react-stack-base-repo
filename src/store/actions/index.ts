import { actionCreatorA } from './actionCreatorA';
import { asyncActionCreatorB } from './actionCreatorB';

export default {
  actionCreatorA,
  asyncActionCreatorB
};
