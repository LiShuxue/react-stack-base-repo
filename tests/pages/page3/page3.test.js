import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Page3 from 'src/pages/page3';
import { actionCreatorA, asyncActionCreatorB } from 'src/store/actions';

jest.mock('react-redux', () => {
  const useSelector = jest.fn();
  useSelector
    .mockImplementationOnce(() => 'this is preload stateA')
    .mockImplementationOnce(() => 'this is preload stateB');

  const dispatch = jest.fn(action => {
    if (typeof action === 'object') {
    }
    if (typeof action === 'function') {
      action();
    }
  });
  const useDispatch = () => dispatch;

  return {
    useSelector,
    useDispatch
  };
});

jest.mock('src/store/actions', () => ({
  actionCreatorA: jest.fn(),
  asyncActionCreatorB: jest.fn()
}));

test('should render page correctly', () => {
  const { asFragment } = render(<Page3 />);
  expect(asFragment()).toMatchSnapshot();
});

test('should have "click this button to change the stateA" button', () => {
  const { getByText } = render(<Page3 />);
  const element = getByText('click this button to change the stateA');
  expect(element).toBeInTheDocument();
});

test('should have "click this button to async change the stateB" button', () => {
  const { getByText } = render(<Page3 />);
  const element = getByText('click this button to async change the stateB');
  expect(element).toBeInTheDocument();
});

test('click first button should dispatch actionCreatorA', () => {
  const { getByText } = render(<Page3 />);
  const element = getByText('click this button to change the stateA');
  fireEvent.click(element);
  expect(actionCreatorA.mock.calls.length).toEqual(1);
});

test('click second button should dispatch asyncActionCreatorB', async () => {
  const { getByText } = render(<Page3 />);
  const element = getByText('click this button to async change the stateB');
  fireEvent.click(element);
  expect(asyncActionCreatorB.mock.calls.length).toEqual(1);
});
