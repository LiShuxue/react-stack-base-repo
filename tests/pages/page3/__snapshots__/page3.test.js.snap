// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render page correctly 1`] = `
<DocumentFragment>
  <div>
    <h1>
      page 3
    </h1>
    <hr />
    <button
      class="ant-btn"
      type="button"
    >
      <span>
        click this button to change the stateA
      </span>
    </button>
    <br />
    <h1>
      stateA: this is preload stateA
    </h1>
    <br />
    <button
      class="ant-btn"
      type="button"
    >
      <span>
        click this button to async change the stateB
      </span>
    </button>
    <br />
    <h1>
      stateB: this is preload stateB
    </h1>
  </div>
</DocumentFragment>
`;
